File: LinkedList/Singly/singly.go
package main

import "fmt"

type LinkedListNode struct {
	value int
	next  *LinkedListNode
}

type SinglyLinkedList struct {
	head   *LinkedListNode
	length int
}

func (l *SinglyLinkedList) pushFirst(value int) {
	previous := l.head
	l.head = &LinkedListNode{value: value}
	l.head.next = previous
	l.length++
}

func (l *SinglyLinkedList) pushLast(value int) {
	node := &LinkedListNode{value: value}
	var current *LinkedListNode

	if l.length == 0 {
		l.head = node
		return
	}

	current = l.head

	for current.next != nil {
		current = current.next
	}

	current.next = node
	l.length++

}

func (l *SinglyLinkedList) pushAt(value, index int) {

	if l.length == 0 {
		l.pushFirst(value)
		return
	}

	if l.length < index {
		return
	}

	node := &LinkedListNode{value: value}
	current := l.head
	var previous *LinkedListNode

	for i := 0; i < index; i++ {
		previous = current
		current = current.next
	}

	node.next = current
	previous.next = node
	l.length++
}

func (l SinglyLinkedList) getFirst() string {
	return fmt.Sprintf("%d", l.head.value)
}

func (l SinglyLinkedList) getAt(index int) string {
	if index < 0 || index >= l.length {
		return ""
	}

	if index == 0 {
		return l.getFirst()

	}

	current := l.head
	for i := 0; i < index; i++ {
		current = current.next
	}

	return fmt.Sprintf("%d", current.value)
}

func (l SinglyLinkedList) getLast() int {
	current := l.head

	for current.next != nil {
		current = current.next
	}

	return current.value
}

func (l *SinglyLinkedList) pop() {
	l.head = l.head.next
	l.length--
}

func (l *SinglyLinkedList) popAt(index int) {
	if index < 0 || index >= l.length {
		return
	}

	if index == 0 {
		l.pop()
		return
	}

	count := 0
	current := l.head
	var previous *LinkedListNode

	for count < index {
		previous = current
		current = current.next
		count++
	}

	previous.next = current.next
	l.length--

}

func (l *SinglyLinkedList) popValue(value int) {

	if l.length == 0 {
		return
	}

	if l.head.value == value {
		l.pop()
		return
	}

	previous := l.head

	for previous.next.value != value {
		if previous.next.next == nil {
			return
		}
		previous = previous.next
	}
	previous.next = previous.next.next
	l.length--
}

func (l *SinglyLinkedList) clear() {
	l.head = nil
	l.length = 0
}

func (l SinglyLinkedList) log() {

	current := l.head
	for l.length != 0 {
		fmt.Printf("%d -> ", current.value)
		current = current.next
		l.length--
	}
	fmt.Printf("null")
	fmt.Println()
}

func main() {
	list := SinglyLinkedList{}

	list.pushFirst(14)
	list.pushFirst(5)
	list.pushFirst(7)
	list.pushFirst(87)
	list.pushFirst(100)
	list.pushFirst(45)
	list.pushLast(90)
	list.pushAt(49, 3)

	// fmt.Println(list.getFirst())
	// fmt.Println(list.getAt(7))
	// fmt.Println(list.getLast())
	list.log()
	// list.popAt(20)
	// list.popValue(45)
	// list.clear()
	// list.pop()
	// list.log()

	// fmt.Println(list)

}

File: LinkedList/Singly/singly.js
class LinkedListNode {
  constructor(value, next) {
    this.value = value;
    this.next = next;
  }
}

/**
 * Implementation of a singly linked list
 */
class SinglyLinkedList {
  constructor() {
    this.head = null;
    this.length = 0;
  }

  pushFirst(value) {
    this.head = new LinkedListNode(value, this.head);
    this.length++;

    return this;
  }

  pushLast(value) {
    let newNode = new LinkedListNode(value);
    let current;

    if (!this.head) {
      this.head = newNode;
    } else {
      current = this.head;

      while (current.next) {
        current = current.next;
      }

      current.next = newNode;
    }
    this.length++;
  }

  pushAt(value, index) {
    if (index < 0 || index > this.length) return;

    if (index === 0) return this.pushFirst(value);

    const newNode = new LinkedListNode(value);
    let prev, curr;

    curr = this.head;
    let count = 0;
    while (count < index) {
      prev = curr;
      count++;
      curr = curr.next;
    }
    newNode.next = curr;
    prev.next = newNode;

    this.length++;
  }
  getFirst() {
    return this.head.value;
  }

  getAt(index) {
    if (index < 0 || index > this.length) return null;
    let current = this.head;

    for (let i = 0; i < index; i++) current = current.next;
    return current.value;
  }

  getLast() {
    let curr = this.head;
    while (curr.next) {
      curr = curr.next;
    }
    return curr.value;
  }

  pop() {
    this.head = this.head.next;
    this.length--;
  }

  popAt(index) {
    if (index < 0 || index > this.length) return null;

    if (index === 0) return this.pop();

    let curr = this.head;
    let prev;
    let count = 0;

    while (count < index) {
      prev = curr;
      curr = curr.next;
      count++;
    }

    prev.next = curr.next;
    this.length--;
  }

  clear() {
    this.head = null;
    this.length = 0;
  }

  log() {
    let output = "";
    let current = this.head;

    while (current) {
      output = `${output}${current.value} -> `;
      current = current.next;
    }
    console.log(`${output}null`);
  }
}

let list = new SinglyLinkedList();

list.pushFirst(14);
list.pushFirst(5);
list.pushFirst(7);
list.pushLast(87);
list.pushAt(100, 3);
list.pushAt(45, 0);
list.log();
// console.log(list.getFirst());
// console.log(list.getLast());
console.log(list.getAt(8));
// list.popAt(1);
// list.log();
